### PR TITLE
indexers: Fix subscribers race.

### LINKED ISSUE #2913: Panic on notifying index subscribers (release 1.7.1)
```
fatal error: concurrent map iteration and map write
```

[Full trace](http://sprunge.us/Y8Xbg5)

https://github.com/decred/dcrd/blob/e1b2e6011bf679a36ed39494dfcb1dcac30bbc80/blockchain/indexers/common.go#L692-L695

Maybe relevant for #2871

### PR DESCRIPTION
**Pull request by dnldd**
This fixes a race accessing an indexer's subscribers waiting on index updates.

fixes  #2913

### CODE REVIEW

File: blockchain/indexers/common.go [hunk 1]
```diff
@@ -460,6 +460,31 @@ func tip(db database.DB, key []byte) (int64, *chainhash.Hash, error) {
...
+// next index update.
+//
+// This should be called with the associated indexer mutex held.
+func fetchIndexerSubscribers(subscribers map[chan bool]struct{}) map[chan bool]struct{} {
+	subs := make(map[chan bool]struct{}, len(subscribers))
+	for subscription := range subscribers {
+		// Remove closed subscriber channels.
+		open := true
+		select {
+		case _, open = <-subscription:
```
[the remark this answers is not in the document]
Comment (dnldd) [reply blockchain/indexers/common.go:473]: good catch, thanks. Will rework it soon.

**Review by davecgh — APPROVED** [review]
EDIT: This still has races and panics.

**Review by davecgh — CHANGES_REQUESTED** [review]
Actually after doing a bit more testing, I see this still isn't fixed and can panic now:

```
panic: close of closed channel

goroutine 71 [running]:
github.com/decred/dcrd/blockchain/v4/indexers.maybeNotifySubscribers({0x18557d0, 0xc000072400}, {0x1857b88, 0xc000075f80})
        src/github.com/decred/dcrd/blockchain/indexers/common.go:727 +0x27e
github.com/decred/dcrd/blockchain/v4/indexers.updateIndex({0x18557d0, 0xc000072400}, {0x1857b88?, 0xc000075f80?}, 0xc037748380)
        src/github.com/decred/dcrd/blockchain/indexers/common.go:813 +0x1a7
github.com/decred/dcrd/blockchain/v4/indexers.(*IndexSubscriber).Run(0xc0001d20c0, {0x18557d0, 0xc000072400})
        src/github.com/decred/dcrd/blockchain/indexers/indexsubscriber.go:360 +0x3f6
main.(*server).Run.func7({0x18557d0?, 0xc000072400?}, 0xc000278780)
        src/github.com/decred/dcrd/server.go:3149 +0x3e
created by main.(*server).Run
        src/github.com/decred/dcrd/server.go:3148 +0x5a5
```

**Comment by dnldd** [comment]
Noticed the issue, `maybeNotifySubscribers` and `purgeIndexerSubscription` are closing the subscriber channel without knowing the state it's in prior. Will need to synchronize the channel closes. Can't see a way around reading the channel to check its state yet, should be fine since the waitForSync receiver is expecting a closed channel when there is an index  update. Looking for a better way however.

**Review by davecgh — APPROVED** [review]
I've performed full syncs on both mainnet and testnet with the race detector enabled and while constantly spamming requests for both `existsaddress` and `getrawtransaction` without issue.

**Review by JoeGruffins — APPROVED** [review]

**Pull request merged**